org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/TryWithResourcesFilter.java

```diff
@@ -0,0 +1,426 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.LabelNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Filters code that is generated for try-with-resources statement.
+ */
+public final class TryWithResourcesFilter implements IFilter {
+
+	public void filter(final MethodNode methodNode,
+			final IFilterOutput output) {
+		if (methodNode.tryCatchBlocks.isEmpty()) {
+			return;
+		}
+		final Matcher matcher = new Matcher(output);
+		for (TryCatchBlockNode t : methodNode.tryCatchBlocks) {
+			if (t.type == null) {
+				matcher.start(t.handler);
+				if (!matcher.matchEcj()) {
+					matcher.start(t.handler);
+					matcher.matchEcjNoFlowOut();
+				}
+			} else if ("java/lang/Throwable".equals(t.type)) {
+				for (Matcher.JavacPattern p : Matcher.JavacPattern.values()) {
+					matcher.start(t.handler);
+					if (matcher.matchJavac(p)) {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	static class Matcher {
+		private final IFilterOutput output;
+
+		private final Map<String, VarInsnNode> vars = new HashMap<String, VarInsnNode>();
+		private final Map<String, String> owners = new HashMap<String, String>();
+		private final Map<String, LabelNode> labels = new HashMap<String, LabelNode>();
+
+		private AbstractInsnNode start;
+		private AbstractInsnNode cursor;
+
+		Matcher(final IFilterOutput output) {
+			this.output = output;
+		}
+
+		private enum JavacPattern {
+			OPTIMAL, FULL, OMITTED_NULL_CHECK, METHOD,
+		}
+
+		private void start(final AbstractInsnNode start) {
+			this.start = start;
+			cursor = start.getPrevious();
+			vars.clear();
+			labels.clear();
+			owners.clear();
+		}
+
+		private boolean matchJavac(final JavacPattern p) {
+			// "catch (Throwable t)"
+			if (!nextIs(Opcodes.ASTORE)) {
```

Comment:
@Godin Is there a specific reason that you don't check the local variable here? The functionality is there and used for other ASTORE/ALOAD checks.